### PR TITLE
Baseball Story

### LINKED ISSUE #122: [Turing Data Story] Baseball Statistics
# Story description 

Given my love of baseball and statistics (my dog-eared copy of *Moneyball* is still a favorite to re-read from time to time) I am interested in writing a story using baseball statistics. Still thinking about the exact details, but it will use the [Lahman Database](http://www.seanlahman.com/baseball-archive/statistics/), so for the moment this is a placeholder to record my interest.

# Ethical guideline 

Ideally a Turing Data Story has these properties and follows the [5 safes framework](http://blog.ukdataservice.ac.uk/access-to-sensitive-data-for-research-the-5-safes/).

- [x] The analysis you produce is openly available and reproducible.
- [x] Any data used are [open](https://opendatahandbook.org/guide/en/what-is-open-data/) and have an explicit licence, provenance and attribution.
- [ ] Any data used are not [personal data](https://ico.org.uk/for-organisations/guide-to-data-protection/guide-to-the-general-data-protection-regulation-gdpr/key-definitions/what-is-personal-data/) (i.e. the data is anonymous or anonymised).
- [x] Any linkage of datasets in your data story does not lead to an increased risk of the personal identification of individuals.
- [x] The Story must be truthful and clear about any limitations of analysis (and potential biases in data).
- [x] The Story will not lead to negative social outcomes, such as (but not limited to) increasing discrimination or injustice.

# Current status

 - [x] Scoping of the story
 - [x] Write story outline
 - [x] Add material to the story notebook
 - [x] Combine materials into a readable story
 - [x] Proofread
 - [x] Request reviews
 - [ ] Address reviews
 - [ ] Merge to master branch.

### PR DESCRIPTION
**Pull request by edaub**
### Summary

Analysis using the Lahman Baseball Database to try and quantify what "replacement level" is for major league hitters using Bayesian Inference. Replacement level refers to some hypothetical player that is cheap and easy to find, and then trying to determine how much better a given player is than this hypothetical one. Replacement level for hitters is usually estimated as 80% of the league average (see for instance [Wikipedia](https://en.wikipedia.org/wiki/Value_over_replacement_player)), but it's unclear what this actually means in precise statistical terms. This story goes into detail using a Bayesian model to quantify this and estimates the level over the course of baseball history.

Fixes #122

Potentially good reviewers might be @jack89roberts or @nbarlowATI 

### List of changes proposed in this PR (pull-request)

* New story using baseball statistics
* Changes to requirements files

### What should a reviewer concentrate their feedback on?

- [ ] Baseball is not as well known in the UK as in the US, Latin America, and Japan. Did I clarify enough about the game and what the statistics that I used mean?
- [ ] Are the SQL queries to get the data not too hard to understand for a novice?
- [ ] Is my approach reasonable? There are a few numbers I tweaked to get the stats to line up well with the standard estimates (percentiles used to quantify a "full season" and how many standard deviations below the average in the final estimate), so if these seem like egregiously unreasonable numbers to use then please let me know.
- [ ] Did I cover the Bayesian methods adequately?
- [ ] Is my coding style reasonable and understandable? I try to write things mostly in functions out of habit, which isn't always straightforward to understand in a literate tool like a notebook.
- [ ] Everything looks ok?

### Acknowledging contributors

- [ ] All contributors to this pull request are already named in the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file.
- [x] The following people should be added to the [table of contributors](https://github.com/alan-turing-institute/the-turing-way/blob/master/README.md#contributors) in the README file: I'd suggest @jack89roberts as a reviewer, so if he gives a review he should be added here.

### CODE REVIEW

**Comment by jack89roberts** [comment]
Nice, I'm happy to take a look through 🙂

**Comment by nbarlowATI** [comment]
I'm also very happy to review!

**Comment by crangelsmith** [comment]
Review issue created (#127) and assigned to @jack89roberts and @nbalowATI.

File: binder/environment.yml [hunk 1]
```diff
@@ -16,3 +16,4 @@ dependencies:
   - autopep8
   - conda-forge::jupyter_contrib_nbextensions
   - pymc3
+  - sklearn
```
Comment (jack89roberts) [inline binder/environment.yml:19]:
```suggestion
  - scikit-learn
```
Binder build was failing with "nothing provides requested sklearn". `pip install sklearn` does seem to work, but docs suggest `pip install scikit-learn` so also suggested changing that in requirements.txt

File: stories/baseball/requirements.txt [hunk 2]
```diff
@@ -0,0 +1,5 @@
+numpy
+scipy
+pandas
+matplotlib
+sklearn
```
Comment (jack89roberts) [inline stories/baseball/requirements.txt:5]:
```suggestion
scikit-learn
```

**Comment by crangelsmith** [comment]
@all-contributors please add @jack89roberts  for review

**Comment by edaub** [comment]
> - Baseball in general
>  - A basic diagram of a baseball field/diamond may be helpful (annotated with terms used in the analysis like S, 2B, 3B, HR where possible)?
>  - The definitions are in the text and the data dictionary, but at times I would have found a glossary of terms helpful, mostly of things like AB, BB, HBP, SH, SF, PA etc. that appear either in queries or data frames (maybe it could go in the data section at the beginning).

>- SQL queries
>  - I think you do a good job of explaining what the queries are doing. The one thing I find a bit tricky is the calculation of fields in `rc_base_query`, e.g. it may not be obvious why `O = AB + SH - SF - H`.

> - Runs created:
>  - I'm not sure I fully understand why this is a good metric/the way it is defined. Naively, I'd expect it to be `TB / 4`, or `TB / 4*PA` if it was a rate stat rather than a counting stat. What's the purpose of the success rate term?
>  - My instinct was that `TB` counts only the number of bases the batter advanced from their initial hits (not bases advanced by teammates at the same time, or bases subsequently gained during teammates' hits). Is that correct? What made me unsure were notes like the ones below, as I'm not sure how they influence `RC` unless `TB` is incorporating more than I thought:
>    - > hits tend to advance any teammates already on the bases a different number of bases
>    - > Extending this to be more accurate requires factoring in base running
>    - > being more precise in how the runs each team scores can be attributed to each player (i.e. the "a player cannot be on base for themselves" effect).
>        - Edit: On reflection, is one nuance that a hitter might have slower teammates on bases in front of them, which limits how many bases they can advance?

> - Did I cover the Bayesian methods adequately?
>  - At the start of "Multinomial Bayesian Interface" is there maybe a more layman's terms way the posterior could be introduced/defined (currently: _"and seek to compute the posterior probability conditioned on the data"_)?
>  - Is it worth discussing why the prior you've selected is reasonable a bit more (or revisiting that later when your results show it was reasonable)?

> - Simulating League Seasons
>  - Could you add a legend to the plot in "Simulating League Seasons" comparing the simulated and empirical distributions?
>  - Theres a sentence that's not finished:
>    - > Others within that peak near zero have not provided evidence that they are good, so these are not useful in evaluating if

> - Gaussian mixture model:
>  - At first I didn't fully comprehend why the GMM was needed; my instinct was you could have just calculated the 2.5% quantile directly from the sample. What I missed is that you're calculating the 2.5% quantile of the second Gaussian only, not the full distribution. Maybe this could be emphasised a bit more in the text?

> - Replacement runs created:
>  - In the summary at the end, perhaps you could comment on how replacement RC could be used by a team, e.g. to identify over-/undervalued players, I think? Maybe out of scope for the story, but could you pick out a player that's much better/worse than their reputation suggests, for example?

> - Is my coding style reasonable and understandable?
>  - In `runs_created` should line 12 `df.ndim == 1` be `season_stats.ndim == 1` (or if not should `df` be an argument to that function rather than relying on using the variable defined earlier in the notebook)?
>  - In `project_player_season` you call an argument `prior` instead of `priors` which is used in other places/functions. I _think_ the `prior` in `project_player_seasons` is identical to the `priors` used elsewhere, so maybe it could be changed to `priors` as well to be consistent?

> - References/links:
>  - It might be nice to add a few more links to wider context around baseball and baseball stats, including to things mentioned in the text like _"Moneyball by Michael Lewis"_, _"legendary baseball statistician Bill James"_, _"dead ball" era prior to 1920"_, _"the late 1990's shows a spike during the 'steriod' era"_, and also to terms related to the analysis like _"Beta distribution"_, _"Dirichlet distribution"_ etc.

**Comment by edaub** [comment]
- Baseball is not as well known in the UK as in the US, Latin America, and Japan. Did I clarify enough about the game and what the statistics that I used mean?

This is tricky...  I can definitely see the risk of giving more detail, and burying the reader in too much information.  However, leaving out some detail does risk things becoming confusing.  For example, I think the difference between "At Bat" and "Plate Appearance" is not really obvious here without looking at the SQL query.  

I would be a bit tempted to go all-in on the basic explanation.  Something like:
 - Every time a batter comes to the plate to face the pitcher, this is a "Plate Appearance"
 - The pitcher will throw the ball, and one of the following will happen:
    - Batter hit by pitch (HBP) - advances to first base.
    - Batter doesn't swing at the ball, ball is outside the "strike zone" - this is a "ball".  Pitcher will throw again. If there are four balls in a Plate Appearance, this is a "Walk" (or "base-on-balls") (BB), batter advances to first base.
    - Batter doesn't swing at the ball, ball is inside the "strike zone", OR the batter swings at the ball and misses - this is a "Strike".  Pitcher will throw again.   If there are three Strikes in a Plate Appearance, the batter is out.
etc. etc.   Could also then mention "batting average", as this is, as far as I can see, the main motivation for having "At Bat" as a separate stat from "Plate Appearance" ?

Maybe this is too much information though - it's definitely a matter of opinion, so I wouldn't argue the case too strongly.

I also didn't quite understand the point about the best players being overvalued because they cannot get on base for themselves...   Is this because although they might get to e.g. 2B, the batters behind them will be worse, so this is less likely to result in a Run?
I would have thought that in this case it's not the level of the player relative to the rest of his team that's important, but relative to the whole league?  (And couldn't this be countered by the fact that the best teams have lots of good players, so good players are more likely to have good team-mates who can bat them in?)

In general though I think the explanations are good, but it might be nice to tell the reader in advance that they are coming - e.g. just before Cell 4, a quick sentence saying that all the acronyms will be explained in the following section.

- [x] Are the SQL queries to get the data not too hard to understand for a novice?
The only tricky one is Cell 5, with the "coalesce" commands - the explanation is good, but I'm torn as to whether this should go before or after the cell itself.    It might also be good to have a sentence here explaining why this command filters out the pitchers - I'm not sure I quite understand that.

- [x] Is my approach reasonable? There are a few numbers I tweaked to get the stats to line up well with the standard estimates (percentiles used to quantify a "full season" and how many standard deviations below the average in the final estimate), so if these seem like egregiously unreasonable numbers to use then please let me know.
- [x] Did I cover the Bayesian methods adequately?
- Yes, really good explanations I think (and useful code snippets!)
- [x] Is my coding style reasonable and understandable? I try to write things mostly in functions out of habit, which isn't always straightforward to understand in a literate tool like a notebook.
- Yes, I think this is excellent, and having everything in functions is useful for any reader who might want to transplant bits and pieces into their own code!
- [x] Everything looks ok?

- I think the plot that is the output of Cell 13, comparing the actual and simulated runs created over a whole season, would benefit from a legend.
- There's a missing half-sentence just after this: "so these are not useful in evaluating if".

**Comment by edaub** [comment]
I just pushed some more revisions based on some changes to the analysis. I think I am happy with this version, though if someone would be able to proofread before publishing I would appreciate it! Thanks.

**Comment by billfinnegan** [comment]
Really great data story - the writing style is clear and accessible, and I like how the intro sets up the concept outside of the details of baseball. Here are a few notes based on a proofread:
- In the Replacement Level section you mention the 80% stat and referred to it on the Wikipedia page for WAR. I scanned the page and couldn't find any reference to replacement being 80% above average - is that based on a calculation of the relative runs in the WAR calc? If so, might be worth explaining, or finding another link, as this rule of thumb is the starting point for your analysis.
- In the Data section, 2nd para, 2nd sentence - I'd change "choose" to "chose". I haven't done a close read of tense, and it might be hard to be consistent (sometimes you are looking back on previous data/analysis, sometimes you are describing what you are doing, other times explaining what you are about to do) but that particular verb jumped out at me.
- End of the data section, you've got two paragraphs starting with "However" which I find a little confusing. Can you drop the first however and start with "Most of the time..."?
- Last para in the data section - I think it might be worth defining "position players" in a brief parenthetical just to make sure that is crystal clear, maybe something like: (the eight defensive fielding positions other than the pitcher). I assume you are ignoring the DH, or at least not getting into that particular quirk. 
- At the beginning of the Runs Created section, I might be getting too caught up in the language, but is "Outs occur when a player fails to hit the ball" totally accurate? "Outs occur when the player fails to get a hit" feels like it might be better, distinguishing between generally making contact and being credited with a hit. Or maybe the slightly longer: "Outs occur when a player fails to hit the ball and advance to the bases."
- Small punctuation thing: at the end of Runs Created, I think "etc)." should be "etc.)." Although, at that point we've just got a HR left, so is that worth mentioning?
- In Runs Created, after the plot, in the bulleted caveats, I think the language in the last bullet is still a little confusing: "because they cannot get on base for themselves" - is it the same to say "because they cannot already be on base while they are batting"?
- In the Bayesian intro, 2nd para, I think you need a comma after "or the probability"
- In Simulating full seasons, before the histogram, missing a word in: "variation in the value this hypothetical player" - of this hypothetical player, or this hypothetical player has?
- In Approx the Dist, first sentence, replace "might we expect" with "we might expect" ?
- In the Mixture Models section, after the plot, when describing the different distributions, can you refer to the labels in the plot?
- In the code block that calcs the estimated and heuristic replacement level, can you also calc the mean runs created/outs so that is handy for understanding the % below league average?
- When you run the model for every year in baseball history, does that start in 1900? The Lahman data looks like it starts in 1871, but I'm not sure when MLB formally began. Might be worth a mention of the start point and logic for that here or when you first reference the data source.
- 2 para below the baseball history plot, 3rd sentence - "Additional thing to consider" probably needs to start with An or One...
- Towards the end, I think it would be good to add a "Discussion" or "Conclusion" header, maybe starting just below the history plot.
- Penultimate paragraph - extra space between their and hits in 3rd sentence.
- Penultimate paragraph - "is" should be "are" or "teams that produce" changed to "a team that produces"
- At the end, I really wanted you to go back to the % below average stat and what the Bayesian model calculates that to be for each year (unless I'm misunderstanding something about how this is working). Is it possible to create a plot of % below average over time as well (again, if I'm understanding this, the heuristic would just be a line at 80% and the bayesian would range from about 60-75%)? Or maybe just in the text for the conclusion you could reiterate the range of runs created/outs and how that reflects % below league average as these are the key #s at the top of the story.

Let me know if you need me to clarify any of this, and feel free to ignore anything!

**Comment by billfinnegan** [comment]
In terms of a one-sentence summary, does this work:
Using Bayesian methods to explore the concept of a "replacement level" baseball player in terms of runs created compared to the league average.

And if we ever want a slightly longer description:
To explore marginal utility - how much additional value you get for paying more for something - you need a baseline, whether you are comparing pastries or professional athletes. In baseball, the baseline is referred to as the replacement level, which is generally thought to be about 80% of the league average. This data story investigates over a century of professional baseball data to calculate the "runs created" by every player, and uses Bayesian methods to calculate the replacement level for each year.

**Comment by edaub** [comment]
Thanks @billfinnegan for the helpful style comments -- I implemented most of them or something in the spirit of what you suggested. I also re-worked a couple of code bits that I thought were inelegant. I think this is ready to publish now!

**Comment by crangelsmith** [comment]
Fantastic @edaub , thank you so much! We will be publishing it on the page in the next week. 

Also, we are working with Comms to advertise the stories on the Turing site and they are asking for a one-sentece summary and a longer description. @billfinnegan has kindly agreed to write it for us:


> In terms of a one-sentence summary, does this work:
> Using Bayesian methods to explore the concept of a "replacement level" baseball player in terms of runs created compared to the league average.
> 
> And if we ever want a slightly longer description:
> To explore marginal utility - how much additional value you get for paying more for something - you need a baseline, whether you are comparing pastries or professional athletes. In baseball, the baseline is referred to as the replacement level, which is generally thought to be about 80% of the league average. This data story investigates over a century of professional baseball data to calculate the "runs created" by every player, and uses Bayesian methods to determine the replacement level for each year.

Are you happy with this summary Eric? If so I'll send it to Comms.
